src/lib/userStorage/UserProfileStorage.js

```diff
@@ -0,0 +1,556 @@
+    this.wallet = wallet
+    this.profiledb = profiledb
+  }
+
+  /**
+   * read raw profile from database
+   */
+  async init() {
+    const rawProfile = await this.profiledb.getProfile()
+    const decryptedProfile = await this._decryptProfileFields(rawProfile)
+    // eslint-disable-next-line array-callback-return
+    Object.keys(this.profileSettings).map(field => {
+      if (!decryptedProfile[field]) {
+        decryptedProfile[field] = {
+          value: null,
+          display: null,
+          privacy: this.profileSettings[field].defaultPrivacy || 'public',
+        }
+      }
+    })
+    this._setLocalProfile(decryptedProfile)
+  }
+
+  /**
+   * helper for setting value of local profile state
+   * @param newValue
+   * @private
+   */
+  _setLocalProfile(newValue: Profile) {
+    this.profile = newValue
+
+    this.onProfileUpdate()
+  }
+
+  onProfileUpdate = debounce(() => over(this.subscribersProfileUpdates)(this.profile), 500, {
+    leading: false,
+    trailing: true,
+  })
+
+  /**
+   * helper for decrypt profile values
+   * @param profile
+   * @returns {Promise<{}>}
+   * @private
+   */
+  async _decryptProfileFields(profile: { [key: string]: ProfileField }): Promise<any> {
+    const outputProfile = {}
+    await Promise.all(
+      Object.keys(profile).map(
+        async item =>
+          typeof profile[item]?.value === 'string' &&
+          (outputProfile[item] = {
+            ...profile[item],
+            value: await this.profiledb.decryptField(profile[item]?.value),
+          }),
+      ),
+    )
+    return outputProfile
+  }
+
+  /**
+   * encrypt decrypted profile
+   * @param {*} profile
+   */
+  async _encryptProfileFields(profile: { [key: string]: ProfileField }): Promise<any> {
+    let encryptProfile = {}
+    await Promise.all(
+      Object.keys(profile).map(async field =>
+        typeof profile[field]?.value === 'string'
+          ? (encryptProfile[field] = {
+              ...profile[field],
+              value: await this.profiledb.encryptField(profile[field]?.value),
+            })
+          : profile[field]?.value === null &&
+            (encryptProfile[field] = {
+              ...profile[field],
+            }),
+      ),
+    )
+    return encryptProfile
+  }
+
+  /**
+   * saves a complete profile to the underlying storage
+   * @param {*} profile
+   * @param update
+   */
+  async setProfile(profile: { [key: string]: string }, update: boolean = false): Promise<any> {
+    if (profile && !profile.validate) {
+      profile = getUserModel(profile)
+    }
+
+    const fields = Object.keys(profile).filter(prop => this.profileSettings[prop])
+    let { errors, isValid } = profile.validate(update)
+
+    if (!isValid) {
+      logger.warn(
+        'setProfile failed',
+        'Fields validation failed',
+        new Error('setProfile failed: Fields validation failed'),
+        { errors, category: ExceptionCategory.Human },
+      )
+
+      throw errors
```

Comment:
please look here: https://github.com/GoodDollar/GoodDAPP/pull/3375#discussion_r685898727

Reply:
(option stuff, could be avoid)